### PR TITLE
Convert competitions to use kagglesdk

### PR DESCRIPTION
**Pull request by stevemessick**
The competitions portion of the API now uses `kagglesdk`.

The HTTP client has ben updated to use prod, if needed.

### CODE REVIEW

File: kaggle/cli.py [hunk 1]
```diff
@@ -1674,11 +1674,11 @@ class Help(object):
     command_model_instances_new = 'Create a new model instance'
     param_model_instance_downfile = (
         'Folder for downloading the special model-instance-metadata.json file '
-        '(https://github.com/Kaggle/kaggle-api/wiki/ModelInstance-Metadata). ')
+        '(https://github.com/Kaggle/kaggle-api/wiki/Model-Metadata#model-instance). ')
```
Comment (rosbo) [inline kaggle/cli.py:1677]: I see a bunch of models changes that seems unrelated to your PR. Can you make sure to rebase it to make it easier to review? TY

Comment (stevemessick) [reply kaggle/cli.py:1677]: `Current branch convert-comps is up to date.`

I had some `git` problems and ended up including commits that have already been merged.

Comment (rosbo) [reply kaggle/cli.py:1677]: Not sure the exact `git` problems you are running into given I don't know your `git` state and which command you ran but your could try the following:

1. Squash all your commits into one: https://stackoverflow.com/questions/5189560/how-do-i-squash-my-last-n-commits-together (this is to make step 2 easier. Simpler to only rebase one commit).
1. Rebase against main: `git rebase origin main`

Comment (stevemessick) [reply kaggle/cli.py]: These changes are no longer included.

File: tests/unit_tests.py [hunk 2]
```diff
@@ -292,6 +292,7 @@ def test_competition_a_list(self):
         try:
             competitions = api.competitions_list()
             self.assertGreater(len(competitions), 0)  # Assuming there should be some competitions
+            api.competitions_list_cli()
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply tests/unit_tests.py:295]: The `_cli` methods do the I/O for the user. They have no return value. If they work, the test passes. The point of calling them is to ensure the new generated objects (from `kagglesdk`) do not break the existing code. I could wrap it in an exception handler, but I don't think that's necessary.

File: kaggle/api/kaggle_api_extended.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/python
 #
-# Copyright 2024 Kaggle Inc
+# Copyright 2019 Kaggle Inc
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py:3]: Actually, this is fixing a previous commit error. We should keep the oldest date.

File: kaggle/api/kaggle_api_extended.py [hunk 4]
```diff
@@ -414,17 +418,17 @@ def authenticate(self):
                 or self.CONFIG_NAME_KEY not in config_data:
             if os.path.exists(self.config):
                 config_data = self.read_config_file(config_data)
-            elif self._is_help_or_version_command(api_command) or (
-                    len(sys.argv) > 2 and api_command.startswith(
-                        self.command_prefixes_allowing_anonymous_access)):
+            elif self._is_help_or_version_command(api_command) or (len(
+                    sys.argv) > 2 and api_command.startswith(
+                self.command_prefixes_allowing_anonymous_access)):
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py:423]: I agree, but ...`yapf3`.

File: kaggle/api/kaggle_api_extended.py [hunk 5]
```diff
@@ -684,6 +688,21 @@ def print_config_values(self, prefix='- '):
         self.print_config_value(self.CONFIG_NAME_PROXY, prefix=prefix)
         self.print_config_value(self.CONFIG_NAME_COMPETITION, prefix=prefix)
 
+    def build_kaggle_client(self):
+        env = KaggleEnv.STAGING if '--staging' in self.args \
+            else KaggleEnv.ADMIN if '--admin' in self.args \
+            else KaggleEnv.LOCAL
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py:694]: I think I need to rewrite the auth code in the HTTP client before we can use `PROD`. I added [WIP] in the title to remind myself not to merge this too soon.

File: kaggle/api/kaggle_api_extended.py [hunk 6]
```diff
@@ -1729,10 +1759,10 @@ def dataset_create_version(self,
                     self.process_response(
                         self.with_retry(
                             self.datasets_create_version_by_id_with_http_info)(
-                                id_no, request)))
+                            id_no, request)))
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py:1765]: Yeah, I don't understand why `yapf3` is changing this, either. I wonder if I accidentally got a newer version. What version do you have?

File: kaggle/api/kaggle_api_extended.py [hunk 7]
```diff
@@ -1953,8 +1983,11 @@ def download_file(self,
         size = int(response.headers['Content-Length'])
         size_read = 0
         open_mode = 'wb'
-        remote_date = datetime.strptime(response.headers['Last-Modified'],
-                                        '%a, %d %b %Y %H:%M:%S %Z')
+        try:
+            remote_date = datetime.strptime(response.headers['Last-Modified'],
+                                            '%a, %d %b %Y %H:%M:%S %Z')
+        except KeyError:
+            remote_date = datetime.now()
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py]: This has been moved to #628.

File: src/kaggle/api/kaggle_api_extended.py
[the remark this answers is not in the document]
Comment (stevemessick) [reply src/kaggle/api/kaggle_api_extended.py:1]: The `kaggle` package is generated from `src/kaggle`, with the addition of all the Swagger-generated code. The `kaggle` package is what gets distributed.

File: kaggle/api/kaggle_api_extended.py [hunk 8]
```diff
@@ -946,10 +971,10 @@ def competition_list_files_cli(self,
         else:
             result = self.competition_list_files(competition, page_token,
                                                  page_size)
-            next_page_token = result.nextPageToken
+            next_page_token = result.next_page_token
             if next_page_token:
                 print('Next Page Token = {}'.format(next_page_token))
-            fields = ['name', 'size', 'creationDate']
+            fields = ['name', 'totalBytes', 'creationDate'] # Breaking change: column header
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py]: n/m -- changed it to preserve `size`. No breaking change!

File: kaggle/api/kaggle_api_extended.py [hunk 9]
```diff
@@ -1195,27 +1232,46 @@ def dataset_list(self,
...
-                                              tagids=tag_ids or '',
-                                              search=search or '',
-                                              user=user or '',
-                                              page=page,
-                                              max_size=max_size,
-                                              min_size=min_size))
-        return [Dataset(d) for d in datasets_list_result]
+            group = DatasetSelectionGroup.DATASET_SELECTION_GROUP_USER
+
+        # datasets_list_result = self.process_response(
```
Comment (jplotts) [inline kaggle/api/kaggle_api_extended.py:1243]: nit - can remove this?

File: tests/unit_tests.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jplotts) [reply tests/unit_tests.py:295]: > If they work, the test passes.

IMO the problem is that "work" really hasn't been verified.  The only thing the test knows is that the server responded with something that can be parsed.  But "something" could be the empty list, so it's not even necessarily checking that the objects themselves can be parsed (which I think is your intention).  The test really needs to verify that at least one competition was returned from the server.  This could be done by refactoring so there's a function that returns the parsed competition list or just by intercepting std out.

**Review by jplotts — COMMENTED** [review]
Thanks!  I'm going to wait to approve until the `prod` flag is working, just to ensure that the PR isn't merged too early.

File: kaggle/api/kaggle_api_extended.py [hunk 10]
```diff
@@ -205,8 +208,8 @@ def _load_previous_if_any(self):
 
     def _is_previous_valid(self, previous):
         return previous.path == self.path and \
-               previous.start_blob_upload_request == self.start_blob_upload_request and \
-               previous.timestamp > time.time() - ResumableFileUpload.RESUMABLE_UPLOAD_EXPIRY_SECONDS
+            previous.start_blob_upload_request == self.start_blob_upload_request and \
```
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py:214]: I still don't know why the formatting changed. I have two versions of `yapf3`. v0.33 is used when running in gLinux and 0.25 is used in the container. It's possible that I had an earlier version in gLinux and it got updated automatically. If I reformat with 0.25 it changes a lot of things, 92 total, but some of those are restoring the change that was present prior to this PR.

File: tests/unit_tests.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stevemessick) [reply tests/unit_tests.py:295]: It sounds like you want the result object to be exercised in the same way it would if the `_cli` method were called, without calling the `_cli` method. I'm lazy, and didn't want to do all that coding. But I guess I can.

[the remark this answers is not in the document]
Comment (stevemessick) [reply tests/unit_tests.py:295]: > The test really needs to verify that at least one competition was returned from the server.

That's done on line 294. Similarly, other tests should have asserts to verify functionality prior to `_cli` being invoked.

File: kaggle/api/kaggle_api_extended.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py]: This can now be run against the production server.

File: tests/unit_tests.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stevemessick) [reply tests/unit_tests.py]: Tests have been updated to show the verification w/o calling the `_cli` version.

File: kaggle/api/kaggle_api_extended.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (stevemessick) [reply kaggle/api/kaggle_api_extended.py]: I switched to `yapf` 0.32, which fixed most of the formatting problems.

File: kaggle/api/kaggle_api_extended.py [hunk 11]
```diff
@@ -716,13 +752,16 @@ def competitions_list(self,
...
-                                                  page=page,
-                                                  search=search or ''))
-        return [Competition(c) for c in competitions_list_result]
+        with self.build_kaggle_client() as kaggle:
+            request = ApiListCompetitionsRequest()
+            request.group = group
+            request.page = page
+            request.category = category
+            request.search = search
+            request.sort_by = sort_by
```
Comment (jplotts) [inline kaggle/api/kaggle_api_extended.py:761]: Just noticing that these have `or` clauses in the original code, e.g. - `category or ''`.  I'm not sure if this is still needed, feel free to ignore if not.

File: src/kaggle/api/kaggle_api_extended.py [hunk 12]
```diff
@@ -332,6 +339,15 @@ class KaggleApi(KaggleApi):
     command_prefixes_allowing_anonymous_access = ('datasets download',
                                                   'datasets files')
 
+    # Attributes
+    competition_fields = ['ref', 'deadline', 'category', 'reward', 'teamCount','userHasEntered']
```
Comment (jplotts) [inline src/kaggle/api/kaggle_api_extended.py:343]: It would be nice to share these with the other file, but I don't think it's critical to refactor right now.

**Review by jplotts — APPROVED** [review]
LGTM, thanks!
